var/spack/repos/builtin/packages/acts/package.py

```diff
@@ -399,7 +399,10 @@ class Acts(CMakePackage, CudaPackage):
     # ACTS imposes requirements on the C++ standard values used by ROOT
     for _cxxstd in _cxxstd_values:
         for _v in _cxxstd:
-            depends_on(f"geant4 cxxstd={_v.value}", when=f"cxxstd={_v.value} {_v.when} +geant4")
+            if Spec(_v.when).satisfies("@0.23:"):  # geant4 variant only exists at 0.23 or higher
+                depends_on(
+                    f"geant4 cxxstd={_v.value}", when=f"cxxstd={_v.value} {_v.when} +geant4"
+                )
```

Comment:
Would this work? It removes the redundancy of having to specify the explicit version here.
```
            if Spec(_v.when).intersects("+geant4"):  # geant4 variant only exists at 0.23 or higher
```

>  I think _probably_ the best thing to do in acts is to just not to use a loop and write out the constraints explicitly

This whole sections could end up looking pretty repetitive (with 23 to be added probably in the next months):
```
    depends_on(f"geant4 cxxstd=17", when="cxxstd=17 +geant4")
    depends_on(f"geant4 cxxstd=20", when="cxxstd=20 +geant4")
    depends_on(f"geant4 cxxstd=14", when="cxxstd=14 +fatras_geant4")
    depends_on(f"geant4 cxxstd=17", when="cxxstd=17 +fatras_geant4")
    depends_on(f"geant4 cxxstd=20", when="cxxstd=20 +fatras_geant4")
    depends_on(f"root cxxstd=14", when="cxxstd=14 +tgeo")
    depends_on(f"root cxxstd=17", when="cxxstd=17 +tgeo")
    depends_on(f"root cxxstd=20", when="cxxstd=20 +tgeo")
```

In hindsight, maybe there is actually no need to specify the `_v.when` part here, as that will come from the variant value contraint. Maybe this is sufficient here:
```suggestion
            depends_on(
                f"geant4 cxxstd={_v.value}", when=f"cxxstd={_v.value} +geant4"
            )
```
That would also remove the never-matching `cxxstd=14 @:0.8.1 +geant4` which #44425 flags, although maybe it also won't like `cxxstd=14 +geant4` (which also never matches anything).

Anyway, this looks good and I am taking too many words to say I approve of the change :-)

Reply:
This suggestion is way better -- it's clearer and it orthogonalizes the constraints.  The solver already knows which versions support which standards from here:

```python
    _cxxstd_values = (
        conditional("14", when="@:0.8.1"),
        conditional("17", when="@:35"),
        conditional("20", when="@24:"),
    )
    _cxxstd_common = {
        "values": _cxxstd_values,
        "multi": False,
        "description": "Use the specified C++ standard when building.",
    }
    variant("cxxstd", default="17", when="@:35", **_cxxstd_common)
    variant("cxxstd", default="20", when="@36:", **_cxxstd_common)
```

Because you have conditional variant values.  Note that the different defaults for different version ranges won't work until #44425 is merged (right now it always defaults to 20) but otherwise the declaration above looks pretty good.

I think you can simplify the whole loop to:

```python
    for _cxxstd in _cxxstd_values:
        for _v in _cxxstd:
            depends_on(f"geant4 cxxstd={_v.value}", when=f"cxxstd={_v.value} +geant4")
            depends_on(f"geant4 cxxstd={_v.value}", when=f"cxxstd={_v.value} +fatras_geant4")
            depends_on(f"root cxxstd={_v.value}", when=f"cxxstd={_v.value} +tgeo")
```

Reply:
Also with this change the strict checks are simplifying, not complicating, in all cases, so that makes me even more happy.

Reply:
Oh and if I can just get back to #41700 once #44425 is done, I think we could also get rid of the loop here.

Reply:
Submitted this separately in https://github.com/spack/spack/pull/46333